### ISSUE TITLE
Provide an image that doesn't require that it be run as root.

### ISSUE THREAD

**Issue by GrahamDumpleton** [DataDog/docker-dd-agent#79]
Your current image requires that it run as root. Because of the potential security issues, it isn't really best practice if there is no need for it to actually run as root.

That your image must run as root means that it is not usable on Docker based cloud infrastructure which restricts users to only running images as a non root user. This is going to make it harder for people to use your product.

That said, if a platform doesn't allow you to run as root, it isn't usually going to allow you to mount the Docker socket either.

The problem at this point is that you are bundling dogstatsd together with the agent for Docker monitoring (having deprecated the old image). The dogstastd should not need to access the Docker socket and so could be used by itself and quite happily run as a non root user. This means it is only of use for collecting application metrics, but when deploying to cloud infrastructure where you don't manage the underlying Docker, this may be all that you want anyway.

So what really should be done is the following:

1. Provide an image where both dogstatsd and the agent can be run as a non root user.
2. If the agent cannot be run as a non root user, then resume providing a separate up to date image for dogstatsd, but ensure that that separate image can be run as a non root user.

As to making dogstatsd runnable as a non root user, it is a simple matter of making certain directories and files group writable. This is sufficient because although platforms will override what user an image runs as, they usually leave the group alone, meaning they run as gid of 0. As the files and directories Docker creates when setting up an image are in the 0 group (root), group write access is adequate.

The files/directories that need group write access are:

```
RUN chmod g+w /etc/dd-agent/datadog.conf
RUN chmod g+w /var/log/datadog
RUN chmod g+w /etc/dd-agent
```

The config file being writable allows ``/entrypoint.sh`` to edit it when run as non root.

The other directories are for logging and socket/lock (???) files.

In addition to the above, there is also one problem deriving from the dogstatsd code that would need to be addressed.

When running just dogstatsd, you do not need the Docker socket for data collection so they do not need to be mounted. The problem is that dogstatsd code when it thinks it is running inside of Docker will try and use the Docker socket to work out what the hostname should be rather than use the internal host name.

The dogstastd code should have a graceful fallback to using the standard UNIX hostname in the container if the Docker socket is not present. Alternatively, this image should accept an environment variable to force dogstatsd to use the system hostname. For example, an environment variable called ``DOGSTATSD_USE_SYS_HOSTNAME``. If this is set, the ``/entrypoint.sh`` script could run:

```
sed -i -e "s/^#hostname:.*$/hostname: `hostname`/" /etc/dd-agent/datadog.conf
```

That is, set the hostname in the configuration file so dogstastd uses that. That way don't have to modify the dogstatsd code.

As to running the agent as a non root user I am not sure what else is required for that. Even with the changes above, trying to running supervisord as non root user just sees it exit straight away and can find nothing logged anywhere indicating why it was failing to start up.

FWIW, you can find my own Docker image, basing off this one, which makes the changes to at least allow dogstatsd to run as a non root user at:

* https://github.com/GrahamDumpleton/testing-docker-dd-dogstatsd

I see this as an interim measure and don't want to have to use it in the future. I would prefer to see DataDog provide an image that can run as non root user.

Also be aware that I intend to write a blog post about using dogstatsd as a sidecar container to an application under OpenShift 3. I will be explaining the current problems as outlined above, in that post. I hope that I will be able to say in that post that DataDog is working on coming out with an updated image that can run as a non root user. If you can it will mean that the DataDog product will be immediately useable to OpenShift 3 users for application monitoring.

**Comment by hkaj** [comment]
Hi @GrahamDumpleton,
First thank you for taking the time to write this detailed report. We really appreciate this kind of feedback!
We are discussing this internally, and will update the issue soon. Stay tuned.

**Comment by remh** [comment]
Thanks again for the great feedback @GrahamDumpleton 

We are finalizing our 5.8.0 agent release that should go out in the upcoming days and we'll work on that feature right after for our 5.9.0 release.

We completely agree that the agent shouldn't run as root or shouldn't need to mount the docker socket when running dogstatsd only. 

Thanks again for the feedback

**Comment by GrahamDumpleton** [comment]
On first test, changes in #97 don't look they result in this working the target environment where have the requirement of running as non root. Need to dig into why. Errors in logs are:

```
sed: couldn't open temporary file /etc/dd-agent/sedg1aKuL: Permission denied
Log file is unwritable: '/var/log/datadog/dogstatsd.log'
2016-09-05 09:24:18,644 | CRITICAL | dd.dogstatsd | utils.dockerutil(dockerutil.py:115) | Unable to find docker host hostname
2016-09-05 09:24:18,741 | CRITICAL | dd.dogstatsd | utils.dockerutil(dockerutil.py:115) | Unable to find docker host hostname
2016-09-05 09:24:18,820 | INFO | dd.dogstatsd | utils.pidfile(pidfile.py:35) | Pid file is: /tmp/dogstatsd.pid
2016-09-05 09:24:18,821 | INFO | dd.dogstatsd | daemon(daemon.py:157) | Starting
2016-09-05 09:24:18,821 | INFO | dd.dogstatsd | daemon(daemon.py:171) | Pidfile: /tmp/dogstatsd.pid
2016-09-05 09:24:18,822 | INFO | dd.dogstatsd | dogstatsd(dogstatsd.py:334) | Listening on host & port: ('', 8125)
2016-09-05 09:24:18,822 | INFO | dd.dogstatsd | dogstatsd(dogstatsd.py:165) | Reporting to https://app.datadoghq.com every 10s
2016-09-05 09:24:28,824 | INFO | dd.dogstatsd | dogstatsd(dogstatsd.py:210) | Flush #1: flushed 0 metrics, 0 events, and 0 service check runs
2016-09-05 09:24:40,204 | ERROR | dd.dogstatsd | dogstatsd(dogstatsd.py:274) | Unable to post payload.
Traceback (most recent call last):
  File "/opt/datadog-agent/agent/dogstatsd.py", line 265, in submit_http
    r.raise_for_status()
  File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/requests/models.py", line 834, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
HTTPError: 403 Client Error: Forbidden
2016-09-05 09:24:40,205 | ERROR | dd.dogstatsd | dogstatsd(dogstatsd.py:276) | Received status code: 403
```

The HTTP error keeps repeating.

**Comment by GrahamDumpleton** [comment]
Not building from ``Dockerfile`` in ``dogstastd`` sub directory. Let me try again.

**Comment by GrahamDumpleton** [comment]
Building image from correct ``Dockerfile`` looks better so far. Need to now test it properly with an application sending data, with ``dogstatsd`` image as a side car container.

```
2016-09-05 09:39:09,126 INFO RPC interface 'supervisor' initialized
2016-09-05 09:39:09,126 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2016-09-05 09:39:09,126 INFO supervisord started with pid 1
2016-09-05 09:39:10,130 INFO spawned: 'forwarder' with pid 10
2016-09-05 09:39:10,131 INFO spawned: 'dogstatsd' with pid 11
2016-09-05 09:39:15,380 INFO success: forwarder entered RUNNING state, process has stayed up for > than 5 seconds (startsecs)
2016-09-05 09:39:15,381 INFO success: dogstatsd entered RUNNING state, process has stayed up for > than 5 seconds (startsecs)
```

Not sure if the ``CRIT`` message is a concern yet or not.

**Comment by hkaj** [comment]
Hi @GrahamDumpleton 
You tested it before the dockerhub build even finished! 😄 
You can find the image under the `11.3.585-dogstatsd` tag on dockerhub, the alpine version will follow shortly.

Sorry for the delay on this one, and please let us know if you have any feedback.

**Comment by GrahamDumpleton** [comment]
I built it myself direct from repo as wasn't sure when the official images would follow on Docker Hub. :-)

**Comment by GrahamDumpleton** [comment]
I am getting data through using your official image using it as a sidecar container. I do have some concerns though how you are using qualified tag for what is providing different functionality than other existing tags for that image. Also need to sort out why your whole image tag naming scheme is getting rejected and can't auto import images. A task for tomorrow.
